### DISCUSSION TITLE
Create server available on HTTP/HTTPs that can be used with PM2

### DISCUSSION THREAD

**Discussion by El-Tommy** [socketio/socket.io#4603] [Q&A] [answered]
I'm trying to create a Socket.IO server that has the following goals:

1. Accessible on the local network of virtual machines using HTTP (`http://<server-local-ip>`)
2. That can be accessed via browser by users through the HTTPs protocol, and that can also make the socket.io.js bundle available via HTTPs (`https://socket-server.example.com`)
3. That uses all available CPUs in the virtual machine (the server will run in just one virtual machine) (Possible with PM2)
4. Have the ability to be automatically restarted in case of failure (Possible with PM2)

For that I created a script based on the Socket.IO help [article](https://socket.io/docs/v4/pm2/) teaching how to use PM2 and this [question](https://github.com/socketio/socket.io/discussions/4600) that teaches to use HTTP and HTTPs.

```
/**
 * pm2 start basic.js -i 0
 */

const http = require("http");
const https = require("https");
const { Server } = require("socket.io");
const { createAdapter } = require("@socket.io/cluster-adapter");
const { setupWorker } = require("@socket.io/sticky");

const { readFileSync } = require("fs");

const httpServer = http.createServer();

const httpsServer = https.createServer({
  key: readFileSync("./localhost-key.pem"),
  cert: readFileSync("./localhost.pem")
});


const io = new Server(httpServer, {
  cors: {
    origin: "*",
    methods: ["GET", "POST"]
  }
});

io.adapter(createAdapter());

setupWorker(io);

io.on("connection", (socket) => {
  console.log(`connect ${socket.id}`);
});

httpsServer.on("request", (req, res) => {
  io.engine.handleRequest(req, res);
});

httpsServer.on("upgrade", (req, socket, head) => {
  io.engine.handleUpgrade(req, socket, head);
});

httpServer.listen(8080);
httpsServer.listen(4430);
```

Using HTTP and HTTPs always throws an error.

Via HTTPs I can't load the socket.io.js bundle. But as this service will be available via browser, it will be necessary to make it available via HTTPs to users.
Direct access via HTTPs displays:

```
{
code: 0,
message: "Transport unknown"
}

```

This is just using the first part of the script, without trying to run with PM2 yet.

When placing the PM2 part next to the script, other errors appear:

- I have to remove the code `httpServer.listen(3000);` for HTTP to work
- When I connect to HTTPs the code never finds the session, so it keeps trying to reconnect endlessly.
- socket.io.js via HTTPs remains unreachable

Even using HTTP socket.io.js and connecting with `<script src="http://localhost:8080/socket.io/socket.io.js"></script>
    <script> const socket = io('https://localhost:3001');</script>` nothing works

However, if I run all this over HTTP only, without requiring HTTPs, it works perfectly.

What am I doing wrong for HTTP/HTTPs not to work together?

Will I have to make the server available only in HTTP and create a proxy via NGINX to support HTTPs and call the HTTP server?

**Comment by darrachequesne** [comment, marked as answer]
(Edited after rereading the question)

> Will I have to make the server available only in HTTP and create a proxy via NGINX to support HTTPs and call the HTTP server?

Yes.

The `@socket.io/pm2` package creates its own HTTP server (plain HTTP, listening on port 8080 by default, see [here](https://github.com/socketio/pm2/commit/8c29a7feb6cbde3c8ef9eb072fee284686f1553f#diff-a2c6de3d3ec7fcd0f209063716ac15eac186c2cf2747fcd92c15ad8fa0ee3fb7R99)), so it won't work with HTTPS.

For future readers, here is the solution without pm2:

```js
import { readFileSync } from "fs";
import * as http from "http";
import * as https from "https";
import { Server } from "socket.io";

const httpServer = http.createServer();

const httpsServer = https.createServer({
  key: readFileSync("./key.pem"),
  cert: readFileSync("./cert.pem")
});

const io = new Server(httpServer);

httpsServer.on("request", (req, res) => {
  io.engine.handleRequest(req, res);
});

httpsServer.on("upgrade", (req, socket, head) => {
  io.engine.handleUpgrade(req, socket, head);
});

io.attachServe(httpsServer);

httpServer.listen(3000)
httpsServer.listen(3001);
```
